### PR TITLE
feat: round robin handover to specific host

### LINKED ISSUE #16982: [CAL-4474] "handover" for round robin
be able to change the organiser of a round robin. alert if the person is not available after submitting and enable them to "handover anyway"

<sub>[CAL-4474](https://linear.app/calcom/issue/CAL-4474/handover-for-round-robin)</sub>

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

- Fixes #16982
- Fixes CAL-4474

 Loom Video: https://www.loom.com/share/43fdf72fef3645f5bc24eae09efb66f4

TODO: Follow up with a PR that shows the users availability status in select with a coloured dot. Green,red etc to ensure that the booking is within their working hours/they're not busy. 

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/features/ee/round-robin/roundRobinManualReassignment.ts
Comment (sean-brydon) [inline packages/features/ee/round-robin/roundRobinManualReassignment.ts:1]: I tried to reuse some logic from the normal roundRobinReassign but ran into type hell - Will make an issue to follow up to tidy this up

File: packages/features/ee/round-robin/roundRobinManualReassignment.ts [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+  if (!newUserHost) {
+    throw new Error(ErrorCode.InvalidRoundRobinHost);
+  }
+
+  if (newUserHost.isFixed) {
+    throw new Error(ErrorCode.UserIsFixed);
+  }
+
+  const originalOrganizer = booking.user;
+  const hasOrganizerChanged = !fixedHost && booking.userId !== newUserId;
```
Comment (CarinaWolli) [inline packages/features/ee/round-robin/roundRobinManualReassignment.ts:101]:
```suggestion
  const hasOrganizerChanged = !fixedHost;
```
Wouldn't this already be enough? If there is no fixed host then organizer would always change as their is always only one RR host assigned

File: packages/features/ee/round-robin/roundRobinManualReassignment.ts [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    // Send cancellation email to original organizer
+    const cancelledEvt = cloneDeep(evt);
+    cancelledEvt.organizer = {
+      email: originalOrganizer.email,
+      name: originalOrganizer.name || "",
+      timeZone: originalOrganizer.timeZone,
+      language: { translate: originalOrganizerT, locale: originalOrganizer.locale || "en" },
+    };
+
+    await sendRoundRobinCancelledEmailsAndSMS(
```
Comment (CarinaWolli) [inline packages/features/ee/round-robin/roundRobinManualReassignment.ts:266]: we also need to send the cancelled email when the organizer didn't change

File: apps/web/components/dialog/ReassignDialog.tsx [hunk 3]
```diff
@@ -39,17 +116,50 @@ export const ReassignDialog = ({ isOpenDialog, setIsOpenDialog, teamId, bookingI
...
+            onValueChange={(val) => {
+              form.setValue("reassignType", val as "team_member" | "round_robin");
+            }}
+            className={classNames("mt-1 flex flex-col gap-4")}>
+            <RadioArea.Item
+              value="round_robin"
+              className={classNames("w-full text-sm")}
+              classNames={{ container: classNames("w-full") }}>
+              <strong className="mb-1 block">{t("round_robin")}</strong>
+              <p>{t("round_robin_reassign_description")}</p>
```
Comment (CarinaWolli) [inline apps/web/components/dialog/ReassignDialog.tsx:130]: translation is missing in common.json for this

File: apps/web/components/dialog/ReassignDialog.tsx [hunk 4]
```diff
@@ -39,17 +116,50 @@ export const ReassignDialog = ({ isOpenDialog, setIsOpenDialog, teamId, bookingI
...
+                }}
+              />
+            </div>
+          )}
+          <DialogFooter>
+            <DialogClose />
+            <Button
+              type="submit"
+              data-testid="rejection-confirm"
+              loading={roundRobinReassignMutation.isPending}>
```
Comment (CarinaWolli) [inline apps/web/components/dialog/ReassignDialog.tsx:158]:
```suggestion
              loading={roundRobinReassignMutation.isPending || roundRobinManualReassignMutation.isPending}>
```

**Review by CarinaWolli — DISMISSED** [review]
I have a booking with the following hosts, with Team Pro Example as the RR host and Team Pro Example 2 as the fixed host:
Screenshot 2024-10-21 at 1 27 19 PM

When reassigning Team Pro Example to Team Free Example, the email to to Team Free Example showed the wrong hosts: 
Screenshot 2024-10-21 at 1 28 29 PM

File: packages/features/ee/round-robin/roundRobinManualReassignment.ts [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+      email: newUser.email,
+      name: newUser.name || "",
+      timeZone: newUser.timeZone,
+      language: { translate: newUserT, locale: newUser.locale || "en" },
+    },
+    attendees: attendeeList,
+    uid: booking.uid,
+    destinationCalendar,
+    team: {
+      members: teamMembers,
```
Comment (CarinaWolli) [inline packages/features/ee/round-robin/roundRobinManualReassignment.ts:207]: I am not sure for what `teamMembers` and `attendeeList` is needed, but why is the logic different than in `roundRobinAssignment`?

Comment (CarinaWolli) [reply packages/features/ee/round-robin/roundRobinManualReassignment.ts:207]: The who section of the updated calendar event is also wrong, this is probably related to team members and `attendeeList`
Screenshot 2024-10-21 at 2 42 19 PM

**Comment by zomars** [comment]
Taking over this PR to get it shipped by EOD. Will address @CarinaWolli 's feedback.

**Comment by zomars** [comment]
Found another issue when reassigning. Fixing.
image

image

image

File: packages/features/ee/round-robin/roundRobinReassignment.ts [hunk 6]
```diff
@@ -160,29 +143,14 @@ export const roundRobinReassignment = async ({
...
-      );
-    }
-  }
+  const teamMembers = await getTeamMembers({
+    eventTypeHosts: eventType.hosts,
+    attendees: booking.attendees,
+    organizer: organizer,
+    previousHost: previousRRHost || null,
+    reassignedHost: reassignedRRHost,
+  });
```
Comment (zomars) [inline packages/features/ee/round-robin/roundRobinReassignment.ts:152]: Something about this change is causing an edge case where the attendeeList is empty. Going to revert this block.

**Comment by zomars** [comment]
> I have a booking with the following hosts, with Team Pro Example as the RR host and Team Pro Example 2 as the fixed host: Screenshot 2024-10-21 at 1 27 19 PM
> 
> When reassigning Team Pro Example to Team Free Example, the email to to Team Free Example showed the wrong hosts: Screenshot 2024-10-21 at 1 28 29 PM

It happens even without fixed hosts.

image

image

image

It still appears the old email

**Review by zomars — APPROVED** [review]
Made some fixes. Approving due to urgency but there's still some polishing to do.

**Pull request merged**